crates/spirv-std/src/lib.rs

```diff
@@ -51,6 +51,26 @@ pub use glam;
 pub use num_traits;
 pub use textures::*;
 
+/// Calls the `OpDemoteToHelperInvocationEXT` instruction, which corresponds to discard() in HLSL
+pub fn demote_to_helper_invocation() {
+    #[cfg(target_arch = "spirv")]
+    unsafe {
+        asm!(
+            "OpExtension \"SPV_EXT_demote_to_helper_invocation\"",
+            "OpCapability DemoteToHelperInvocationEXT",
+            "OpDemoteToHelperInvocationEXT"
+        );
+    }
+}
+
+/// Calls the `OpKill` instruction, which corresponds to discard() in GLSL
+pub fn discard() {
+    #[cfg(target_arch = "spirv")]
+    unsafe {
+        asm!("OpKill", "%unused = OpLabel");
```

Comment:
Hah, I didn't think of this when we needed `discard()` for a shadertoy - I'm guessing this just bypasses `rspirv::dr`'s representation of blocks but still results in correct SPIR-V output? I wonder if it has a risk of breaking some of the `linker` algorithms, tho that won't happen if it gets serialized and reparsed.

Reply:
haha yeah it's extremely cursed. only reason it works is because we serialize to object files and re-parse in the linker.